### ISSUE TITLE
[RFC] Remove `type` from Catalog Locations

### ISSUE THREAD

**Issue by Rugvip** [spotify/backstage#2604]
**Status:** Open for comments

## Need

Catalog locations are quite awkward to work with, as they are heavily tied to the processing in the catalog, even in places where you wouldn't expect the user to have to care about such things. For example, when registering a new component in the catalog, you have to know what type of location to select. There's an "Auto-detect" option, but that's backed by a somewhat awkward lookup table [here](https://github.com/spotify/backstage/blob/95fb7b5e4e59e05861c1deb4052132bc75ed9899/plugins/register-component/src/components/RegisterComponentPage/RegisterComponentPage.tsx#L89), which can only take us so far.

There are also cases where it's very desirable or even required to be able to specify a location with a single string, for example when it is used in annotation values. Then you end up with something like [this](https://github.com/spotify/backstage/blob/95fb7b5e4e59e05861c1deb4052132bc75ed9899/docs/features/software-catalog/well-known-annotations.md#backstageiomanaged-by-location), or [that](https://github.com/spotify/backstage/blob/95fb7b5e4e59e05861c1deb4052132bc75ed9899/docs/features/software-catalog/well-known-annotations.md#backstageiodefinition-at-location), or [this](https://github.com/spotify/backstage/blob/95fb7b5e4e59e05861c1deb4052132bc75ed9899/docs/features/software-catalog/well-known-annotations.md#backstageiotechdocs-ref).

## Proposal

Remove `type` from locations and location entities, and instead use the schema as the initial switch for what type of location it is. For example, `http` and `https` for many of our existing processors, or `ldap` for reading ldap data, as well as possible custom protocols like `github-org:github.com/spotify` or `catalog-rest:https://example.com/v1/catalog-integration`.

Within that initial protocol switch, we can then implement per-protocol methods for further refining the selection, for example, for `http` and `https` we would select reader based on the hostname as well. For example `github.com` would default to our GitHub reader, but it would also be possible to register custom readers as well, such as another GitHub reader for `github.internal.example.com`. We may or may not fall back to a plain HTTP fetch if no specific reader is found as well.

Don't wanna dive into implementation in this RFC, just discuss if it's a change that we want to do.

## Alternatives

Keeping location types, but trying to solve some of the UX issues around them in other ways. For example we could provide an endpoint on the catalog that can tell what location type a give URL might be, or a client-side API for the same.

## Risks

I view the management of the different custom schemas as a bit of a risk, but tbh that isn't really related to this move as we'd have to manage the location types otherwise anyway.

One risk is confusion around how to read HTTP sources is custom ways, and confusion around how to signal if it's HTTP or HTTPS. I brought up the example `catalog-rest:https://example.com/v1/catalog-integration`, which could be a custom ingestion mechanism for catalog entities over REST. We're a bit back to square 1 here though, since we have the duplicate schema in the URI. We could do something like `catalog-rest:example.com/v1/catalog-integration`, but would that be HTTPS or not?

We may also need some kind of awareness of the context that a reader is being used in, for example we wouldn't want to allow a `github-org` reader that is used for reading organizational data into the catalog, to all of a sudden be usable to read API schema files. Could be a flag on the reader that specifies that kind of thing it reads, e.g. `'entities' | 'data'` or smth.

**Comment by freben** [comment]
My mental model is that processors would just be matching on string prefix, and that's that. :) What prefixes you match may be part of your config, or not - the github processor may be configured to match on both `https://github.com` and `https://git.example.com`, but the ldap one maybe only ever matches on `ldap:` and has a forced hostname as part of its "contract" for the string syntax. If you don't match, you return false and let somebody else try to match. I don't mind it sometimes looking like a double protocol; the string isn't meant to be parseable as a URI always, and end users will most likely only ever interact with plain https links to their SCM of choice.

**Comment by Rugvip** [comment]
@freben Yeah plain prefix hopefully works just fine too. Bit wary of not fully matching the hostname in case of `example.com.bad.site.com` kinda things, but as long as we can handle it properly it'll simplify the implementation and configuration quite a bit.

**Comment by hooloovooo** [comment]
If we're going this route, wouldn't it make sense to remove the prefix entirely and let the annotation name specify what it is?

**Comment by freben** [comment]
Bear in mind that these are the locations that you add into the catalog, to tell it where to get entities. So it's the URL that you as a user paste into the Backstage interface to register a new entity. The suggestion is to just send that plain to the backend, without trying to infer some higher level type of location before doing so, and the backend processors should just know how to deal with them.

**Comment by andrewthauer** [comment]
Seems cleaner to me if it is handled on the backend. If a user is manually registering a URL from the UI would they also specify the location protocol? Or would that be omitted and inferred though the scheme and/or host?

**Comment by freben** [comment]
The latter. The idea here is to remove that complexity that you speak of and let the backend be smarter in knowing how to handle things.äs

**Comment by Rugvip** [comment]
TL;DR: Let's keep `type`, but consolidate most of out different readers into a `url` type.

Discussed this a bit more on Discord. We're thinking that a good middle ground would be to keep the `type` in the `Location` entity `spec`, and instead add a separate concept for generic data fetching from URLs, basically something that implements `(url: string) => Promise<Buffer>`. That facility would then be used for fetching data for locations of the type `url`, which would end up replacing all of the existing URL-like readers like `github`, `gitlab`, etc. With that in place we can assume that all locations registered by users are of the type `url`, but we can still provide a bit of flexibility in the configuration of different integrations like `github-org`.

The separate URL fetching would probably use the hostname-switch based approach for selecting the concrete logic for doing fetching for any given URL, similar to what we discussed above.

So for the 3 annotations mentioned as examples in the RFC:
- `backstage.io/managed-by-location`: This is replaced by a relationship, see #1964
- `backstage.io/definition-at-location`: Would use the URL reading to read the data
- `backstage.io/techdocs-ref`: Not covered by this solution. An idea would be to add an optional `readTree` method to URL reader interface, which would be able to read an entire file tree from a URL.

**Comment by freben** [comment]
`backstage.io/definition-at-location` may also be supplanted by #2628, which in turn will use the mentioned URL reader
